### DISCUSSION TITLE
Get the line of code committed in each language by a user

### DISCUSSION THREAD

**Discussion by termjobTest** [community/community#24510] [Apps, API and Webhooks]
<p>In my java application, I have a use case of counting the line of code committed by each git user in each language (eg: Java - 1000 LOC, CPP 500 LOC .etc). I am planned to follow the below steps.</p>
<ol>
<li>GET all the commits in each repo by the user as an author:<br>
<a href="https://api.github.com/repos/%7BrepoOwnerId%7D/%7BrepoId%7D/commits?author=%7BauthorId%7D" rel="nofollow noopener">https://api.github.com/repos/{repoOwnerId}/{repoId}/commits?author={authorId}</a>
</li>
<li>GET the commit details <a href="https://api.github.com/repos/%7BrepoOwnerId%7D/%7BrepoId%7D/commits/%7BcommitId%7D" rel="nofollow noopener">https://api.github.com/repos/{repoOwnerId}/{repoId}/commits/{commitId}</a><br>
In the response, we have the total LOC and file name, I am planning to identify the language in committed file based on the file extension (eg: committed file end with .java then its LOC count for JAVA language)</li>
</ol>
<p>But performance-wise I am feeling which is a bad one. Also, Git Api allows only 5000 hits per hour for an authenticated user. Could anyone please suggest any better option for the same.</p>

**Comment by francisfuzz** [comment, marked as answer]
<aside class="quote no-group" data-username="termjobTest" data-post="1" data-topic="121256">
<div class="title">
<div class="quote-controls"></div>
 termjobTest:</div>
<blockquote>
<p>But performance-wise I am feeling which is a bad one. Also, Git Api allows only 5000 hits per hour for an authenticated user. Could anyone please suggest any better option for the same.</p>
</blockquote>
</aside>
<p>:wave: welcome to the GitHub Support Community, <a class="mention" href="/termjobtest">@termjobTest</a> ! Thanks for asking your question here.</p>
<p>As far as I’m aware, there’s not one particular endpoint nor is there a set of endpoints beyond what you mentioned that would give you the lines of code committed (added or deleted) per user. While you could use the GitHub API to collect that information, you can obtain the same information by cloning the repository and using <code>git</code>'s built-in commands to compute the data that you’re looking for. This approach avoids making any calls to the GitHub API.</p>
<p>Here’s an example (though I imagine there are more ways that others in our community can offer input on).</p>
<p>Clone and change directory into <a href="https://github.com/termjobTest/test">one of your repositories</a>:</p>
<pre><code class="lang-auto">git clone git@github.com:termjobTest/test.git 
cd test
</code></pre>
<p>There’s a <a href="https://git-scm.com/docs/git-diff-tree"><code>git-diff-tree</code></a> command that compares the content and mode of blobs found via two tree objects. Run it:</p>
<pre><code class="lang-auto"># Shows the diff-tree for the latest HEAD.
git diff-tree --format=short --numstat -r HEAD 

Author: ajit-indiemaker &lt;53966513+ajit-indiemaker@users.noreply.github.com&gt;

    Create ajittest3.rb

105	0	ajittest3.rb
</code></pre>
<p>It’s possible to format the output to show specific metadata, like the commit SHA, the author information, and even the files changed including how many lines were added and deleted. Here’s the documentation for each of the flags used:</p>
<ul>
<li><a href="https://git-scm.com/docs/git-diff-tree#Documentation/git-diff-tree.txt---formatltformatgt"><code>--format</code></a></li>
<li><a href="https://git-scm.com/docs/git-diff-tree#Documentation/git-diff-tree.txt---numstat"><code>--numstat</code></a></li>
<li><a href="https://git-scm.com/docs/git-diff-tree#Documentation/git-diff-tree.txt--r"><code>-r</code></a></li>
</ul>
<p>You could then write a script that iterates over all of the commits on a single branch (and possibly iterate through all branches) performing this command. <code>--format</code> is configurable so you can choose what kind of output you’d like to see for your use case. I hope this helps!</p>
